### ISSUE TITLE
Questions about self.predict_epsilon and Maze2D settings

### ISSUE THREAD

**Issue by xmlyqing00** [jannerm/diffuser#6]
Hi,

Nice paper and code. They're very clear and easy to understand. I have two questions about the code implementation:

  1. Although I've read articles about the background of the Diffusion model, I'm still new to it. I understand the loss function of comparing epsilons and using p process (the U-Net model) to estimate noise, but I'm confused why you set self.predict_epsilon == False in the locomotion example. That's amazing. What's the loss function and logic behind it? Could you briefly explain or refer me to somewhere to read please? 
  
  2. I think we can use the same model for path planning in Maze 2D, but I'm not sure about the code implementation. Please correct me if I'm wrong,
     a. At initialization, the points are in order which represent the visiting order, and their 2D positions are random.
     b. The constraints are (1) the first/last points should be at the start/end position, (2) the path between consecutive should not hit/cross the walls (I  don't know how to implement constraint yet).
     c. Both training and testing stage, we use the same function `apply_condition` to apply the constraints.

Thanks for your time. Look forward to your response.

Best

**Comment by jannerm** [comment]
1. Section 3.2 of the [DDPM paper](https://arxiv.org/abs/2006.11239) has a great discussion of the different prediction targets and how they are related. In particular:

> To summarize, we can train the reverse process mean function approximator $\mu_\theta$ to predict $\tilde{\mu}_t$, or by modifying its parameterization, we can train it to predict $\epsilon$. (There is also the possibility of predicting $x_0$, but we found this to lead to worse sample quality early in our experiments.) 

We found that the choice between predicting $\epsilon$ (the added noise) and $x_0$ (the noiseless data) to be less important for trajectories than for images.

2. **(a)** At initialization, $x^N$ is Gaussian noise of dimensionality `[ horizon x (state_dim + action_dim) ]`. You could possibly say that the points already have an order because of their position in $x^N$, but I'm not sure it's meaningful to say "$x_0^N$ precedes $x_1^N$", since they are both just noise anyway. **(b)** We don't enforce that collision constraint explicitly; the samples respect this by virtue of there being no data that walks through the walls. **(c)** Yep!

**Comment by xmlyqing00** [comment]
Thanks for your detailed and quick response! Appreciate!!

1. I missed this comment in their paper, thanks for pointing out.
2. I see. For maze planning, I know `horizon=128/256/384` (from your Appendix). Can we say `state_dim=2` for 2D positions? I'm confused about the definition of action and I don't know which one is better or you used in your experiments, (a) direction to the next point, (b) indices to the next point, (c) positions to the next move, or neither of them LOL. I don't know.
3. As for the reward part, please correct me if I am wrong:
    a. $O_t = 1$ means current status $(s_t, a_t)$ is valid (not in the wall).
    b. $p=(\tau | O_{1:T} = 1)$ means this sampled trajectory $\tau$ is valid at all moments. And this plan (trajectory) will be executed from the first action.
    c. In Maze 2D $\mathcal{J}_\phi$ is not a network but a deterministic function that returns the reward of $\mathcal{J}(\mu)$ and its gradient during training. I'm confused about how to compute the gradients $\nabla{s_t,a_t} r(s_t,a_t)$ where $(s_t, a_t)=\mu_t$ here. Does $r$ return either 0 or 1 here? I'm not sure what does `no data` mean?

Sorry that I put more questions here. These are the problems when I try to implement your work in Maze 2D. My main confusions are about `action` and `reward`. Really appreciate if you can take some time to answer my questions.

Cheers

**Comment by xmlyqing00** [comment]
When I read the d4rl dataset about the Maze2d settings, I have some answers to the above questions but have some confusions now.

1. It seems that the diffusion model learns the correct paths in the maze during training, which implicitly encodes the walls. In the evaluation time, it should be the same maze map. The diffusion model can estimate the correct path between two given points.
2. The `conditions` of the maze setting is only the begin and end locations (maybe velocity). The trajectory is `state+action`, where state is $(x,y,v_x,v_y)$, and action is force $(x, y)$ by the D4RL settings. I'm curious how the diffusion model guarantees the `observation` + `action`  equals to the `next observation`.

Thanks.

**Comment by yilundu** [comment]
Hi,

The diffusion model implicitly learns from data the bias that the next observation is equivalent to the previous observation and action.

The optimality condition is that both that the dynamics is valid as well as the reward is optimal. The reward function is neural network that is separately trained on rewards collected from the data. We use the gradient of that network to bias planning. Feel free to let us know if you have any other questions

**Comment by xmlyqing00** [comment]
Thanks @yilundu . I see. The reward function using neural network makes sense.  I think the input to reward function is a trajectory `horizon * (state_dim + act_dim)`, and the target is the reward. The network architecture could be a regression network like MLP.

1. How do we prepare the samples for reward network training? Directly from the D4RL maze dataset, or adding some negative samples like cross the walls.
2. It's the same to my last questions, do the diffusion network and rewards network associate/bias to one maze map after training? If I have a new maze map, do they need to retrain to fit the new map?

Thanks for your answers. Appreciate.

**Comment by xmlyqing00** [comment]
I see. Thanks for your answers! Very insightful!

**Comment by erdisayar** [comment]
> 1. The samples for the reward network training is the same D4RL maze dataset
> 2. Yes, you would need to retrain the model on a new map (although if a map changes only slightly, you can add test time constraints to the map).

Does it mean that you get the reward values from the D4RL maze dataset? Because I couldn't see any reward function or neural network calculating the reward value in the diffuser repo.

There is also a question here (https://github.com/jannerm/diffuser/issues/7).

Have you already updated the repo?

Thanks for your answers in advance
